crates/teloxide/src/dispatching/tracing.rs

```diff
@@ -0,0 +1,34 @@
+use super::{DpHandlerDescription, UpdateHandler};
+
+use dptree::{di::Injectable, prelude::DependencyMap, HandlerDescription};
+use tracing::{Instrument, Span};
+
+/// Extension methods for adding `tracing` to [`UpdateHandler`].
+///
+/// This trait is only available if the `tracing` feature is enabled.
+pub trait UpdateHandlerTracingExt<E> {
+    /// Returns an `UpdateHandler` with tracing enabled.
+    fn with_tracing_span<F, FnArgs>(self, f: F) -> Self
+    where
+        F: Injectable<DependencyMap, Span, FnArgs> + Send + Sync + Clone + 'static;
+}
+
+impl<E: 'static> UpdateHandlerTracingExt<E> for UpdateHandler<E> {
+    fn with_tracing_span<F, FnArgs>(self, f: F) -> UpdateHandler<E>
+    where
+        F: Injectable<DependencyMap, Span, FnArgs> + Send + Sync + Clone + 'static,
+    {
+        // FIXME: This is a hacky replacement for `handler.description().clone()`.
+        // Ideally cloning `DpHandlerDescription` would be supported by `teloxide`.
+        let description = DpHandlerDescription::entry().merge_chain(self.description());
+
+        dptree::from_fn_with_description(description, move |deps: DependencyMap, cont| {
+            let self_c = self.clone();
+            let f_c = f.clone();
```

Comment:
I don't think you need this `.clone()` (or the `Clone` bound), `inject` already take a `&`.

Reply:
Yeah this is basically why I'm doing it:

![image](https://github.com/teloxide/teloxide/assets/21188659/ac7f85af-4dd8-428d-b0a8-5ca521779cf8)

I couldn't find another way to solve this.

Reply:
Try doing
```diff
-let span = f.inject(&deps)().await;
+let f = f.inject(&deps);
+let span = f().await;
+drop(f);
```

Cause I think this should be equivalent to what other dptree things do: https://github.com/teloxide/dptree/blob/da3de9d583380cfe2192f10d1d7681c3e79339a1/src/handler/map.rs#L83-L85

Reply:
I see, there you're using an `Arc` to allow `Fn` to capture the variable. Makes sense, this works now, added the changes to the PR.